modules/shared/src/main/scala/org/tessellation/schema/balance.scala

```diff
@@ -57,6 +66,34 @@ object balance {
         }
     )
 
+    def applyTransactions[F[_]: Applicative: MonadCancelThrow](
+      transactions: Set[Transaction],
+      getBalanceFn: Address => F[Balance]
+    ): F[Map[Address, Balance]] = {
+      val sources = transactions.groupBy(_.source).keySet
+      val destinations = transactions.groupBy(_.destination).keySet
+      val addresses = sources ++ destinations
+
+      def applyTransactions(
```

Comment:
It iterates over the whole transaction list for each balance so complexity could be improved probably